sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/arithmetic.scala

```diff
@@ -0,0 +1,491 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import scala.math.{max, min}
+
+import ai.rapids.cudf._
+import ai.rapids.cudf.ast.BinaryOperator
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.shims.{GpuTypeShims, SparkShimImpl}
+
+import org.apache.spark.sql.catalyst.analysis.{DecimalPrecision, TypeCheckResult}
+import org.apache.spark.sql.catalyst.expressions.{Expression, NullIntolerant}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.shims.RapidsErrorUtils
+import org.apache.spark.sql.types._
+
+abstract class CudfBinaryArithmetic extends CudfBinaryOperator with NullIntolerant {
+
+  protected def allowPrecisionLoss = SQLConf.get.decimalOperationsAllowPrecisionLoss
+
+  // arithmetic operations can overflow and throw exceptions in ANSI mode
+  override def hasSideEffects: Boolean = super.hasSideEffects || SQLConf.get.ansiEnabled
+
+  override def dataType: DataType = (left.dataType, right.dataType) match {
+    case (DecimalType.Fixed(p1, s1), DecimalType.Fixed(p2, s2)) =>
+      resultDecimalType(p1, s1, p2, s2)
+    case _ => left.dataType
+  }
+
+  protected def resultDecimalType(p1: Int, s1: Int, p2: Int, s2: Int): DecimalType = {
+    throw new IllegalStateException(
+      s"${getClass.getSimpleName} must override `resultDecimalType`.")
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = (left.dataType, right.dataType) match {
+    case (l: DecimalType, r: DecimalType) if inputType.acceptsType(l) && inputType.acceptsType(r) =>
+      // We allow decimal type inputs with different precision and scale, and use special formulas
+      // to calculate the result precision and scale.
+      TypeCheckResult.TypeCheckSuccess
+    case _ => super.checkInputDataTypes()
+  }
+}
+
+object GpuAdd extends GpuAddParent {
+  // Scala requires the companion objects to be in the same file
+}
+
+case class GpuAdd(
+    left: Expression,
+    right: Expression,
+    failOnError: Boolean) extends CudfBinaryArithmetic {
+  override def inputType: AbstractDataType = TypeCollection.NumericAndInterval
+
+  override def symbol: String = "+"
+
+  override def binaryOp: BinaryOp = BinaryOp.ADD
+
+  override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.ADD)
+
+  override def doColumnar(lhs: BinaryOperable, rhs: BinaryOperable): ColumnVector = {
+    val ret = super.doColumnar(lhs, rhs)
```

Comment:
I thought that we needed to do the add as something larger than a 128-bit value. Is that not true? Do we just not have tests that cover these odd use cases?  Like add Decimal(30, 0) to a Decimal(30, 10)? This also goes for subtract too...

Reply:
I will fix that as a separate PR

Reply:
Then please add a comment explaining that even if this does fix the current test failures this still has a bug in it and point to that issue.